### PR TITLE
article about local stateful

### PR DESCRIPTION
**Pull request by tgermain**
Signed-off-by: Timothée GERMAIN <timothee.germain@corp.ovh.com>

### CODE REVIEW

File: en/_posts/2016-07-04-local-stateful.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+layout: post
+title:  "Local stateful"
+categories: docker
+author: tgermain
+lang: en
+---
+
+
+
+This tutorial explains how local stateful and volumes work.
```
Comment (yadutaf) [inline en/_posts/2016-07-04-local-stateful.md]: What do you mean by "local stateful and volumes" ? Is it "Locally stateful containers and volumes ?"

File: en/_posts/2016-07-04-local-stateful.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+Here we create a directory `~/data/toto` on the host, let's call it the **hostPath**. We create a file *hello* in it.
+
+When we run the container, we tell docker to map `~/data/toto` aka the **hostPath** to `/data/`, the **containerPath**, inside the container.
+
+The binding is bidirectionnal : if the container creates a file `/data/fromTheOtherSide`, we are able to see it from the host.
+
+
+
+To protect our infrastructure, we restrain what a user can mount into its containers.
```
Comment (yadutaf) [inline en/_posts/2016-07-04-local-stateful.md]: I'd rephrase it like, "There are some limitations on what you can mount in your container..." This puts the focus on the user rather than the technical/security constraints of the infrastructure.

File: en/_posts/2016-07-04-local-stateful.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+The **hostPath** is modified from `/etc/localtime` to `/containerdata/user-toto/etc/localtime`.
+
+
+This path does not exist on the host, when the container starts docker creates it as a directory, it's the default behaviour.
+If the code inside the container tries to read  `/etc/localtime` (the **containerPath**) as a file, it fails because it's a directory on the host.
+
+
+
+If you run into that kind of problem, you can try to troubleshoot it in the "Debug" tab of your faulty application:
```
Comment (yadutaf) [inline en/_posts/2016-07-04-local-stateful.md]: - "you can troubleshoot". People usually don't want to try, but to succeed  ;)
- "of your application"

File: en/_posts/2016-07-04-local-stateful.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+```
+Message executor
+
+    failed to run container: docker failed to start container  0cf43b13cbd46912d43df6f760a21d5fb28e759ecd74ed240d156fef96d34347: Error  response from daemon: rpc error: code = 2 desc = "oci runtime error:  could not synchronise with container process: not a directory"
+```
+
+The message is cryptic but show that a directory was treated as a file and it didn't go well.
```
Comment (yadutaf) [inline en/_posts/2016-07-04-local-stateful.md]: "This message basically means..." "cryptic" tends to suggest mystery hence to discourage from learning.

File: en/_posts/2016-07-04-local-stateful.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+
+## User namespace remap
+The local storage feature currently has issues with permissions. Currently, we are only able to deal with data owning if the process inside the container runs as root.
+
+Why ? Because we use usernamespace remap to isolate containers. It map the user *root* (uid 0) inside the container to an unprivileged user on the host (uid 65536 for example).
```
Comment (yadutaf) [inline en/_posts/2016-07-04-local-stateful.md]: This is hugely technical, I'd suggest to split this part into an overview and a technical details paragraph. It could look like:

> If you intend to use the local storage feature, make sure to start your application as *root*. This is a known limitation and we are working on removing it.
> Under the hood, we use "[User namespace](https://blog.yadutaf.fr/2016/04/14/docker-for-your-users-introducing-user-namespace/)". They allow us to provide you with a root user in your container which is not the actual Host's root. While this improves the application security, this also comes with some constraints  ...

Note: there area a couple of missing "s" on the 3rd person. Like "It map*s*", "which use*s*", ...

**Comment by tgermain** [comment]
PR updated, thanks for your feedback.

File: en/_posts/2016-07-04-local-stateful.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
+---
+layout: post
+title:  "Local stateful containers"
```
Comment (yadutaf) [inline en/_posts/2016-07-04-local-stateful.md]: Maybe you wanted to update this as well ?

Comment (tgermain) [inline en/_posts/2016-07-04-local-stateful.md]: Indeed, better be consistent.
